From: Oleg Nesterov <oleg@redhat.com>
Subject: getrusage: use __for_each_thread()
Date: Sat, 9 Sep 2023 19:26:29 +0200

do/while_each_thread should be avoided when possible.

Plus this change allows to avoid lock_task_sighand(), we can use rcu
and/or sig->stats_lock instead.

Link: https://lkml.kernel.org/r/20230909172629.GA20454@redhat.com
Signed-off-by: Oleg Nesterov <oleg@redhat.com>
Cc: Eric W. Biederman <ebiederm@xmission.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 kernel/sys.c |    4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

--- a/kernel/sys.c~getrusage-use-__for_each_thread
+++ a/kernel/sys.c
@@ -1830,10 +1830,8 @@ void getrusage(struct task_struct *p, in
 		r->ru_oublock += sig->oublock;
 		if (maxrss < sig->maxrss)
 			maxrss = sig->maxrss;
-		t = p;
-		do {
+		__for_each_thread(sig, t)
 			accumulate_thread_rusage(t, r);
-		} while_each_thread(p, t);
 		break;
 
 	default:
_
